### DISCUSSION TITLE
a Jax convolv2d for real 2D arrays

### DISCUSSION THREAD

**Discussion by jecampagne** [jax-ml/jax#15204] [Q&A]
Well, 
I was quite surprise to not find an equivalent of scipy `fftconvolve` in the current JAX version. I probably miss something.
So I have setup a personal crude implementation [nb in Colab ](https://colab.research.google.com/drive/1Ze7nxlQQc05FT_bt1YuxX7PjjYBmBe17?usp=sharing)

I was wandering is my `convolve2r` function is compliant with jax.vmap & jax.jacrev. Of course any suggestion to speed up and make better implementation is welcome. One thing for instance is that the scipy fft `next_fast_len` function is not implemented in JAX. This function compute the size in power of small numbers (2,3,..) greater to the original size, to get FFT working well.
```python
def convolve2r(in1,in2):

  # suppose
  # in1 = jnp.asarray(x1)
  # in2 = jnp.asarray(x2)
  #  "assert" in1.ndim == in2.ndim == 2
  #  "assert" in1.size == in2.size
  #  "assert" in1.dtype.kind == 'f' and in2.dtype.kind == "f"

  axes = range(in1.ndim)

  s1 = in1.shape
  s2 = in2.shape
  shape = [s1[i] + s2[i] -1 for i in range(in1.ndim)]
  
  # crude next_fast_len equivalent with just power of 2
  fshape = 2**jnp.ceil(jnp.log2(jnp.array(shape))).astype('uint16')

  fslice = tuple([slice(sz) for sz in shape])

  sp1 = jnp.fft.fftn(in1, fshape, axes=axes)
  sp2 = jnp.fft.fftn(in2, fshape, axes=axes)
  conv = jnp.fft.irfftn(sp1 * sp2, fshape, axes=axes)
  conv = conv[tuple(map(slice, shape))]   #conv = jax.lax.slice(conv,(0,0),(shape))

  def _centered(arr, newshape):
    # Return the center newshape portion of the array.
    newshape = jnp.asarray(newshape)
    currshape = jnp.array(arr.shape)
    startind = (currshape - newshape) // 2
    endind = startind + newshape
    myslice = [slice(startind[k], endind[k]) for k in range(len(endind))]
    return arr[tuple(myslice)]

  return _centered(conv, s1)
```

Thanks

**Comment by jakevdp** [comment]
Thanks for the question and the demo code!

As written, I don't think this would be compatible with JAX transforms like `jit` and `vmap` because your slice start and end indices are dynamic. I think you could probably fix that by using `np.asarray` in place of `jnp.asarray` within the `_centered` helper function.

The best way to test it is to try it out: put `@jax.jit` on the line before your function and see if it works 😁

I think we'd be open to adding this to the `jax.scipy` package if you want to clean it up and open a PR.

**Reply by jecampagne** [reply]
Hi @jakevdp I was not aware that I can use "pure" numpy (in _centered) because I was wandering how JAX will trace when one want to compute jacrev jacobian wrt the argument of convolv2dr?
For the PR, of course I would be glad to do it.

**Reply by jecampagne** [reply]
BTW jit is ok if I apply your advise and also change jnp -> np here too

```python
fshape = 2**np.ceil(np.log2(np.array(shape))).astype('uint16')
```
Now I am wandering the vmap & jacrev. At least `jax.jacrev(convolve2r)(gal,psf)` crashes on Calob

**Reply by jakevdp** [reply]
You can use pure numpy on trace-time constants, in order to create new trace-time constants. Shapes and slices need to be trace-time constants, because JAX does not support dynamic shapes, so using pure numpy is not only possible here, it is required in order to work with JAX transformations like jit and vmap (using JNP operations on ints in autodiff is fine because it keeps integer values concrete)

**Reply by jecampagne** [reply]
Ok! now what about jacrev crash?

**Reply by jakevdp** [reply]
It works for me if I change `jnp` to `np` for operations involving shapes and indices. If you can be more specific about the code you're running when you see a crash, I may be able to offer better advice.

**Reply by jecampagne** [reply]
@jakevdp have you access to the nb I mentionne in the first post?
The crash is a "Run time restart"

**Reply by jakevdp** [reply]
Thanks - I think it's a memory issue. The output of `jacrev` is the input size times the output size, so for input size `(128,128)` and output `(128,128)` the jacobian is size `(128,128,128,128)` which in float64 is 2GB. Because of the way you define your function, an intermediate result is of size `(256**2, 256, 256)`, which takes 32GB in float64.

Is the fully-instantiated 4D jacobian really something you need to compute? If so, you'll probably have to do it with a smaller problem, or find a larger machine.

**Reply by jecampagne** [reply]
Thanks for the question. Let me draw a schema of possible use of this convolve2d function.

1. Let us define a set of parameters Theta of interest (eg. cosmological  params, and others)
2. Imagine that we compute from Theta a 2D map of image distortion : s = f_s(Theta) NxN array
3. Consider now a 2D map of a galaxy profile g_p  distorted by "s": x_g = f_g(g_p, s)  NxN array  
4. Consider a point spread function response of a telescope as a 2D Map : x_psf NxN array
5. To get a simulated 2D image of the galaxy viewed by the telescope we do a convolution: \tilde{x}_g = convol(x_g,x_psf) NxN image
6. Now we have an real galaxy image, and we make a loss function : loss( \tilde{x}_g(Theta), x_g^{mes})   : scalar
7. To get the best Theta that minimize the loss, we need to perform a gradient descent and compute the \frac{\partial loss}{\partial Theta}

Of course one can argue many things, but let us think of such scenario, what the gradient at step 7 will generate as amount of memory?

**Reply by jecampagne** [reply]
Making an update of the notebook with a simple exercise where I only consider as Theta parameters  the flux and sigma of the galaxy and fix the psf map (and I do not consider distortion)

```python
def model(par, psf):
  flux, sig = par
  gal = f(xx,yy,s=sig)
  gal = flux * gal/sum(gal)
  gal = convolve2r(gal,psf)
  return gal

def loss(par, img_target, psf):
  img_pred = model(par, psf)
  return jnp.mean(jnp.square(img_pred - img_target))  
```
Then
```python
par_init = jnp.array([1e5,1.])
loss(par_init, gal_mes,psf) # Array(37.1882973, dtype=float64)
jax.grad(loss)(par_init, gal_mes,psf) # Array([ 1.24994667e-03, -1.00068581e+02], dtype=float64)
```
Is ok

And using `jaxopt`
```python
gd = jaxopt.GradientDescent(loss, maxiter=1000)
init_params = jnp.array([1.e5,1.])
res = gd.run(init_params,img_target=gal_mes, psf=psf)
```
one gets the minimizer converging nicely.

So, I was wandering how the gradient process does not see the bottleneck of the convolve2dr computation. I would appreciate a nice explanation.

**Reply by jakevdp** [reply]
Performing gradient descent does not involve storing the full Jacobean matrix, so I wouldn’t expect you to run into the same memory issues as when you call jacrev directly - that’s why I asked if your goal is really to compute the full Jacobean matrix.

**Reply by jecampagne** [reply]
So the grad wrt the 2 shape parameters is cleverly traceback to not having to compute the full jacobian as intermediate step, isn't it?

**Reply by jecampagne** [reply]
Ok, the last thing is the vmap: considering several galaxy images that all should be convolved either with the same PSF or with 1 PSF per galaxy.

**Reply by jecampagne** [reply]
Well, here is a vmap test
```python
def raw_generator(par,xx,yy):
  flux, sig = par
  gal = f(xx,yy,s=sig)
  gal = flux * gal/jnp.sum(gal)
  return gal

#make a catalog of raw galaxy images (here 2)
catalog_raw = jnp.array([raw_generator((flux,sig), xx,yy) for flux,sig in zip((0.5e5,1.e5),(1.5,2.))])
#apply pSf on each galaxy image thanks to vmap
catalog=jax.vmap(convolve2r, in_axes=(0,None))(catalog_raw, psf)
```
Seems to give some relevant results. So, I can make a PR
